tests/unit/shared-heap/shared_heap_test.cc

```diff
@@ -92,149 +94,956 @@ destroy_module_env(struct ret_env module_env)
     EXPECT_EQ(1, argv[0]);
 
-    // test aot
-    test_shared_heap(shared_heap, "test.aot", "test_malloc_fail", 1, argv);
+    test_shared_heap(shared_heap, "test.aot", "test_malloc_fail", 0, argv);
     EXPECT_EQ(1, argv[0]);
+
+    test_shared_heap(shared_heap, "test_chain.aot", "test_malloc_fail", 0,
+                     argv);
+    EXPECT_EQ(1, argv[0]);
+}
+
+TEST_F(shared_heap_test, test_preallocated_shared_heap_malloc_fail)
+{
+    SharedHeapInitArgs args = { 0 };
+    WASMSharedHeap *shared_heap = nullptr;
+    uint32 argv[1] = { 0 }, BUF_SIZE = os_getpagesize();
+    uint8 preallocated_buf[BUF_SIZE];
+
+    /* create a preallocated shared heap */
+    args.pre_allocated_addr = preallocated_buf;
+    args.size = BUF_SIZE;
+    shared_heap = wasm_runtime_create_shared_heap(&args);
+    if (!shared_heap) {
+        FAIL() << "Create preallocated shared heap failed.\n";
+    }
+
+    /* test wasm can't malloc with preallocated shared heap */
+    argv[0] = 1024;
+    test_shared_heap(shared_heap, "test.wasm", "my_shared_heap_malloc", 1,
+                     argv);
+    EXPECT_EQ(0, argv[0]);
+
+    argv[0] = 1024;
+    test_shared_heap(shared_heap, "test.aot", "my_shared_heap_malloc", 1, argv);
+    EXPECT_EQ(0, argv[0]);
+
+    argv[0] = 1024;
+    test_shared_heap(shared_heap, "test_chain.aot", "my_shared_heap_malloc", 1,
+                     argv);
+    EXPECT_EQ(0, argv[0]);
+}
+
+static void
+create_test_shared_heap(uint8 *preallocated_buf, size_t size,
+                        WASMSharedHeap **shared_heap_res)
+{
+    SharedHeapInitArgs args = { 0 };
+    WASMSharedHeap *shared_heap = nullptr;
+    args.pre_allocated_addr = preallocated_buf;
+    args.size = size;
+    shared_heap = wasm_runtime_create_shared_heap(&args);
+    if (!shared_heap) {
+        FAIL() << "Create preallocated shared heap failed.\n";
+    }
+
+    *shared_heap_res = shared_heap;
+    if (!*shared_heap_res) {
+        FAIL() << "Create shared heap chain failed.\n";
+    }
+}
+
+static void
+create_test_shared_heap_chain(uint8 *preallocated_buf, size_t size,
+                              uint8 *preallocated_buf2, size_t size2,
+                              WASMSharedHeap **shared_heap_chain)
+{
+    SharedHeapInitArgs args = { 0 };
+    WASMSharedHeap *shared_heap = nullptr, *shared_heap2 = nullptr;
+    args.pre_allocated_addr = preallocated_buf;
+    args.size = size;
+    shared_heap = wasm_runtime_create_shared_heap(&args);
+    if (!shared_heap) {
+        FAIL() << "Create preallocated shared heap failed.\n";
+    }
+
+    memset(&args, 0, sizeof(args));
+    args.pre_allocated_addr = preallocated_buf2;
+    args.size = size2;
+    shared_heap2 = wasm_runtime_create_shared_heap(&args);
+    if (!shared_heap2) {
+        FAIL() << "Create preallocated shared heap failed.\n";
+    }
+
+    *shared_heap_chain =
+        wasm_runtime_chain_shared_heaps(shared_heap, shared_heap2);
```

Comment:
Who is going to unchain and release this shared_heap_chain at the end of every case(if the case `create_test_shared_heap_chain()`)?

Reply:
no need to unchain, runtime will destroy the shared heap list and free the resources